### ISSUE TITLE
The napari_get_writer hookspec should get more info about the layers

### ISSUE THREAD

**Issue by jni** [napari/napari#1637]
## 🚀 Feature

At a minimum, the data from `_get_state` would be really useful. For example, some plugins might only be able to save luminance, not RGB, data. Others might not be able to save a multiscale layer but they can save single scale. etc.

## Alternatives

Currently, the only option for the plugin is to fail late, during the actual writing. This seems suboptimal!

**Comment by GenevieveBuckley** [comment]
I'll also suggest that `_get_state` should also return the layer type.

This is less interesting right now, but when we have layergroups, this becomes more useful to include. (Eg: then the layergroup `_get_state` can easily be a simple combination of all the individual layers state information without a lot of extra special handling to include what type each layer in the group is)

EDIT: wait sorry, this suggestion is in the wrong place - I misread this issue as talking about changes in `_get_state` itself instead of to the writer hookspec. Oops. I also agree with Juan's original suggestion.

**Comment by DragaDoncila** [comment]
@sofroniewn @tlambert03 Thoughts on this? It's quite timely now I would say, particularly with the alfa cohort hopefully producing their own plugins (albeit maybe not writers specifically). 

For example I'm in the process of writing a writer which needs to know the shapes of the given layers in order to determine whether it can write them or not. It doesn't seem like this should be left to the writer function, which would then immediately fail.

**Comment by tlambert03** [comment]
It all seems reasonable to me!

**Comment by sofroniewn** [comment]
I'm very open to this. I'd want to think about how to do it in a backwards compatible way if possible.

Right now we have:

```python
def napari_get_writer(path: str, layer_types: List[str])
```

We could layer done

```python
def napari_get_writer(path: str, layer_meta: List[dict])
```

(where `layer_type` or `type` should probably be a kwarg in the `dict`). Could just call it `meta` too. This might be tricker to make backwards compatible.

I do wonder though then if we should just move away from the `get_writer` approach to just having a function that can write data. The `get_writer` idea came from wanting to mirror the `get_reader` approach which @tlambert03 wanted to be very fast, but if I remember correctly @tlambert03 you did say that we might not want to take this approach for writing too (I believe I was the one who suggested it though).

Then we could just do something like 

```python
def napari_write_layers(path: str, data: List[Any], meta:List[dict])
```
and after #2066
```python
def napari_write_viewer(path: str, viewer:dict)
```
where we are passing the JSON serialization of the `viewer`.
We could then wait to have
```python
def napari_write_layers(path: str, layers:dict)
```
where layers is the JSON serialization of our layers after we finish our pydantic refactor.

Not sure how others feel about the urgency here, or if there is scope creep with my suggestions, but we might want to wait until finishing the pydantic refactor and we have good JSON serialization before adding/ changing these hookspecs. Very open to other ideas though

**Comment by tlambert03** [comment]
just want to point out: there is a mechanism in place to let hookspecs add new arguments over time without breaking old ones. hook implementations [may accept _less_ arguments than defined in the spec](https://pluggy.readthedocs.io/en/latest/#opt-in-arguments).  So this proposal doesn't necessarily need to change `layer_types` to `layer_meta` and deal with complicated backwards compatibility.  Plugins can opt-in to the new information just be accepting more args

_edit_: this, by the way, is one of the reasons why hookspecs can't use arbitrary kwargs... as mentioned in https://github.com/napari/napari/issues/2278#issuecomment-779918184

**Comment by psobolewskiPhD** [comment]
@DragaDoncila is this still relevant or relevant-ish?

**Comment by DragaDoncila** [comment]
@psobolewskiPhD i think this issue as it stands can be closed. The writers *do* receive information from the `get_state` function.

I think there's probably still scope for a discussion on giving even more to writers (I'm struggling to remember the exact issue I ran into recently), but that should be a separate issue, if it arises.
